Refactor WordBoard Initialization and Attributes

yilmazcabuk opened the pull request:
## Description
This pull request refactors the `WordBoard` class initialization process and modifies its attributes for improved readability and maintainability. The changes also introduce dictionaries for various attributes and update the calculation of board attributes based on the new configuration.

## Changes
- Modified the initialization process of the `WordBoard` class for improved clarity.
- Changed the `words` attribute to be a set instead of a list.
- Introduced dictionaries for `letter_values,` `letter_multipliers,` `word_multipliers`, and `board_value`.
- Updated `word_values` to be a list of tuples containing word values and words.
- Renamed the parameter of the `set_board` method to `game_board`.
- Recalculated various attributes based on the updated board configuration.

## Motivation
The motivation behind these changes is to enhance the organization and maintainability of the `WordBoard` class. By introducing dictionaries and renaming attributes, we aim to make the code more understandable and easier to extend in the future.

yilmazcabuk committed:
refactor: improve WordBoard initialization and attributes

Refactor the initialization process of the WordBoard class and modify its attributes.
- Change the 'words' attribute to be a set instead of a list.
- Introduce dictionaries for 'letter_values', 'letter_multipliers', 'word_multipliers', and 'board_value'.
- Update 'word_values' to be a list of tuples containing word values and words.
- Modify 'set_board' method parameter name to 'game_board'.
- Calculate various attributes based on the new board configuration.

yilmazcabuk committed:
Merge branch 'vscala:main' into main

vscala commented on spellcast.py:
```diff
@@ -88,6 +97,19 @@ def __init__(self):
+        self.total_count = Counter()
+        self.board_value = {}
+        self.word_values = []
+        self.calculate_value = None
```
It looks like `calculate_value` is only used in the recalculate function so it probably doesn't need to be a attribute of WordBoard.

yilmazcabuk replied:
Done.

vscala commented on spellcast.py:
```diff
@@ -88,6 +97,19 @@ def __init__(self):
+        self.row_count = 0
+        self.column_count = 0
+        self.total_count = Counter()
+        self.board_value = {}
```
It looks like `board_value` is never actually used, it can probably be removed.

yilmazcabuk replied:
Removed.

vscala commented on spellcast.py:
```diff
@@ -88,6 +97,19 @@ def __init__(self):
         managing the game board.
         """
+        self.words = set()
+        self.letter_values = {}
```
`letter_values` can probably be moved to a local scope in recalculate.

yilmazcabuk replied:
Scoped.

vscala commented on spellcast.py:
```diff
@@ -88,6 +97,19 @@ def __init__(self):
         Reads words from a file and initializes various attributes needed for
         managing the game board.
         """
+        self.words = set()
```
`words` is being initialized here as a set but becomes a list on line 114, consider making this a list or changing line 114 to  a set.

yilmazcabuk replied:
Fixed and renamed.

yilmazcabuk committed:
refactor: reorganize and optimize WordBoard methods

This commit refactors and optimizes the methods in the WordBoard class. Changes include:

- Renaming the 'words' attribute to 'words_set' for clarity.
- Combining attribute initialization into single dictionary comprehensions.
- Moving the 'calculate_value' function definition outside the class.
- Reorganizing and cleaning up the backtrack function for readability.
- Simplifying word and letter value calculations using comprehensions and built-in functions.
- Removing unnecessary comments and white spaces.

vscala commented on spellcast.py:
```diff
@@ -130,35 +124,32 @@ def recalculate(self):
-        self.word_values = [(self.calculate_value(word), word) for word in self.words]
+
+        def calculate_value(word):
+            sum(
```
This is missing `return`

yilmazcabuk replied:
It was actually working as is but still the suggested change was applied.

yilmazcabuk committed:
refactor: optimize word value calculation

Refactor the `calculate_value` function in the `WordBoard` class for better readability and maintainability.

Instead of using `sum()` without assignment, now the values are accumulated in the `value` variable. This change enhances the code's clarity.

Additionally, the conditional expression for applying bonus points to long words has been reformatted to improve code organization.
